Get view definition for SQL Server

kitloong opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | -

#### Summary

Currently, `SQLServer2012Platform::getListViewsSQL` does not export view definition.
This PR updated the SQL so that `Doctrine\DBAL\Schema\View::getSql()` will return view definition instead of empty string.

morozov commented on src/Platforms/SQLServer2012Platform.php:
```diff
@@ -1049,7 +1049,9 @@ public function getCreateViewSQL($name, $sql)
     {
-        return "SELECT name FROM sysobjects WHERE type = 'V' ORDER BY name";
+        return "SELECT name, definition FROM sysobjects
+                    INNER JOIN sys.sql_modules ON (sysobjects.id = sys.sql_modules.object_id)
```
Looks like parentheses are redundant here.

morozov reviewed (commented):
Thank you for the patch. Please add an integration test or reuse an existing one.

morozov commented:
Given that the DBAL doesn't use views itself, I'd rather classify this as an improvement. In this case, the patch should target `3.2.x`.

kitloong commented:
Hi @morozov , thank you for review.

I will add tests for this.

I am sorry, should I change the current PR base to `3.2.x`? Or checkout a new branch from `3.2.x` to commit the changes?

morozov commented:
> […] should I change the current PR base to `3.2.x`?

Yes, this was the first step. Now you need to rebase your changes on `3.2.x` and force-push the changes:
```shell
$ git checkout fix-sqlserver-view
$ git rebase HEAD~ --onto 3.2.x # this assumes you want to rebase one commit
$ git push -f # adjust according to your remote naming
```

kitloong commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -692,7 +708,8 @@ public function testCreateAndListViews(): void
 
-        self::assertTrue($this->hasElementWithName($this->schemaManager->listViews(), $name));
+        self::assertTrue($this->hasElementWithName($views = $this->schemaManager->listViews(), $name));
+        self::assertTrue($this->viewSqlContainsTableName($views, 'view_test_table'));
```
Reusing `testCreateAndListViews`.

Added new assertion to check if `View::getSql()` always contains `view_test_table`, this also cover `SQLServerPlatform`.

morozov replied:
This assertion will make sure that at exactly one of the view definitions contains `view_test_table`. We need to make sure that specifically, the definition with the given name contains the SQL in question.

I assume that some platforms will expand `SELECT *` to an explicit set of columns. Would it make sense to change the query to something like `SELECT id ...` and assert that the entire query is available as part of the definition?

kitloong replied:
> some platforms will expand SELECT * to an explicit set of columns

Thank you for this information. Has changed to `SELECT id ...`  in the new commit.

The tricky part is the different platforms return different format definitions.
For instance:
```
# MySQL
CREATE VIEW doctrine_test_view AS (select `db`.`view_test_table`.`id` AS `id` from `db`.`view_test_table`)

# PgSQL
CREATE VIEW doctrine_test_view AS SELECT view_test_table.id FROM view_test_table

# SQLServer
SELECT id FROM view_test_table
```

The new test a000089 assert `View::getSql()` contains SQL string with regex pattern`/\b(select|id|from|view_test_table)\b/i`
Should be better than the previous test.

morozov replied:
Thanks for the details. In this case, the regex-based assertion looks too fragile, and it's better to just assert that the definition has the table name in it.

kitloong replied:
Thank you, has been updated to assert the definition contains the table name.

Extra note, `$views` key is view name, and again different platform has own format.
```
# MySQL
doctrine_test_view

# PgSQL
public.doctrine_test_view

# IBM DB2
DOCTRINE_TEST_VIEW
```

Has to use `View::getName()` and `strtolower` on key to pass test

kitloong commented:
Hello @morozov 

Following items completed.

- [x] Code update based on last reviewed.
- [x] Rebase 3.2.x.
- [x] Update integration test.

Please review.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -120,6 +124,18 @@ static function (AbstractAsset $item) use ($name): bool {
         return count($filteredList) === 1;
     }
 
+    private function viewContainsSql(View $view, string $sql): bool
```
This logic is not reused anywhere in the test. Please revert to the original assertion and inline into the test method.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -686,13 +702,14 @@ public function testCreateAndListViews(): void
         $this->schemaManager->dropAndCreateView($view);
 
-        self::assertTrue($this->hasElementWithName($this->schemaManager->listViews(), $name));
+        self::assertTrue($this->hasElementWithName($views = $this->schemaManager->listViews(), $name));
```
Assignment in the method call is unreadable. Please extract it into a separate expression.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -110,14 +110,24 @@ public function testDropAndCreateSequence(): void
+     *
+     * @return AbstractAsset[]
+     */
+    private function filterElementByName(array $items, string $name): array
```
Should it be `filterElementsByName()`?

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -686,13 +696,20 @@ public function testCreateAndListViews(): void
 
         $name = 'doctrine_test_view';
-        $sql  = 'SELECT * FROM view_test_table';
+        $sql  = 'SELECT id FROM view_test_table';
```
This change is no longer necessary.

morozov commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -686,13 +696,20 @@ public function testCreateAndListViews(): void
-        self::assertTrue($this->hasElementWithName($this->schemaManager->listViews(), $name));
+        $views = $this->schemaManager->listViews();
+
+        self::assertTrue($this->hasElementWithName($views, $name));
```
The two following lines perform the same assertion. Is this one needed?

morozov commented:
@kitloong looks good. Please squash again.

derrabus commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -692,7 +702,12 @@ public function testCreateAndListViews(): void
+        $views = $this->schemaManager->listViews();
+
+        $filtered = array_values($this->filterElementsByName($views, $name));
+        self::assertSame(1, count($filtered));
```
```suggestion
        self::assertCount(1, $filtered);
```

kitloong committed:
Get view definition for SQL Server

Currently, SQLServer2012Platform::getListViewsSQL does not export view definition.
This commit get view definition for SQL server.
Reference: https://docs.microsoft.com/en-us/sql/relational-databases/views/get-information-about-a-view?view=sql-server-ver15#to-get-the-definition-and-properties-of-a-view

morozov commented:
Thanks, @kitloong!
